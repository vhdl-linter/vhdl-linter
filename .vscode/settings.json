{
  "typescript.tsdk": "node_modules/typescript/lib",
  "eslint.lintTask.enable": true,
  "eslint.lintTask.options": "--ext .ts .",
  "eslint.format.enable": true,
  "cSpell.words": [
    "actuals",
    "clientrc",
    "downto",
    "Inoutput",
    "Mentionables",
    "nand",
    "Onig",
    "OSVVM",
    "quickfix",
    "Referencable",
    "setq",
    "subarchitecture",
    "Textmate",
    "ulogic",
    "uninstantiated",
    "Uniquify",
    "Verilog",
    "vhdl",
    "Vhdl",
    "vmode",
    "vprop",
    "vunit",
    "Whitespaces",
    "xnor"
  ]
}