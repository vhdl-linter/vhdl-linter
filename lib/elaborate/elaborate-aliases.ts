import { implementsIHasReference } from "../parser/interfaces";
import { OAlias, OFile, ParserError } from "../parser/objects";

export function elaborateAliases(file: OFile) {
  for (const alias of file.objectList) {
    if (alias instanceof OAlias) {
      if (alias.name.length === 0) {
        throw new ParserError(`Alias without name`, alias.range); // TODO: change the parser error to addMessage
      }
      alias.aliasDefinitions = alias.name[0].definitions;
      for (const read of alias.name[0].definitions) {
        if (implementsIHasReference(read)) {
          read.aliasReferences.push(alias);
        }
      }
    }
  }

}