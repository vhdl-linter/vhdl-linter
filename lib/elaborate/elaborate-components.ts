import { OFile, OPackage } from "../parser/objects";
import { ProjectParser } from "../project-parser";
import { getEntities } from "./elaborate-instantiations";

export function elaborateComponents(file: OFile, projectParser: ProjectParser) {
  for (const architecture of [...file.architectures, ...file.packages.filter(p => p instanceof OPackage) as OPackage[]]) {
    for (const component of architecture.components) {
      component.definitions.push(...getEntities(component, projectParser));
      const entityPorts = component.definitions.flatMap(ent => ent.ports);
      for (const port of component.ports) {
        port.definitions.push(...entityPorts.filter(p => p.lexerTokenEquals(port)));
      }
      const entityGenerics = component.definitions.flatMap(ent => ent.generics);
      for (const generics of component.generics) {
        generics.definitions.push(...entityGenerics.filter(g => g.lexerTokenEquals(generics)));
      }
    }
  }

}