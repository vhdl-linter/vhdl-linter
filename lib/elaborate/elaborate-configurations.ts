import { OConfiguration, OFile } from "../parser/objects";
import { ProjectParser } from "../project-parser";

export function elaborateConfigurations(file: OFile, projectParser: ProjectParser) {
  for (const configuration of file.objectList) {
    if (configuration instanceof OConfiguration) {
      // find project entities
      configuration.definitions.push(...projectParser.entities.filter(e => e.lexerToken.getLText() === configuration.entityName.getLText()));
      for (const configurations of configuration.definitions) {
        configurations.referenceConfigurations.push(configuration);
      }
    }
  }
}
