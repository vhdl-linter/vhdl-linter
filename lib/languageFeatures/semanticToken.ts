import { SemanticTokenModifiers, SemanticTokens, SemanticTokensBuilder, SemanticTokensLegend, SemanticTokenTypes } from "vscode-languageserver";
import * as I from "../parser/interfaces";
import * as O from "../parser/objects";
import { VhdlLinter } from "../vhdlLinter";
export const semanticTokensLegend: SemanticTokensLegend = {
  tokenTypes: Object.values(SemanticTokenTypes),
  tokenModifiers: Object.values(SemanticTokenModifiers)
};
function tokenType(type: SemanticTokenTypes) {
  return semanticTokensLegend.tokenTypes.indexOf(type);
}

function tokenModifier(modifiers: SemanticTokenModifiers[]) {
  let result = 0;
  for (const modifier of modifiers) {
    const index = semanticTokensLegend.tokenModifiers.indexOf(modifier);
    if (index > 0) {
      result |= 1 << index;
    }
  }
  return result;
}
type BuilderParams = [number, number, number, number, number];
function pushToken(buffer: BuilderParams[], range: O.OIRange, type: SemanticTokenTypes, modifiers: SemanticTokenModifiers[]) {
  const line = range.start.line;
  const char = range.start.character;
  const length = range.end.i - range.start.i;
  // Ignore tokens without length (this is only some implicit declared libraries etc.)
  if (length === 0) {
    return;
  }
  buffer.push([line, char, length, tokenType(type), tokenModifier(modifiers)]);
}


function findDefinition(obj: O.ObjectBase) {
  if (I.implementsIHasDefinitions(obj) && obj.definitions.length > 0) {
    return obj.definitions[0];
  }
}

function pushCorrectToken(buffer: BuilderParams[], obj: O.ObjectBase, definition: O.ObjectBase, range: O.OIRange, fixedModifiers: SemanticTokenModifiers[], colorInputs: boolean) {
  if (definition instanceof O.OArchitecture || definition instanceof O.OEntity || definition instanceof O.OPackage
    || definition instanceof O.OPackageBody || definition instanceof O.OConfigurationDeclaration || definition instanceof O.OLibrary) {
    pushToken(buffer, range, SemanticTokenTypes.class, [...fixedModifiers]);
  } else if (definition instanceof O.OEnum) {
    pushToken(buffer, range, SemanticTokenTypes.enum, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.ORecord) {
    pushToken(buffer, range, SemanticTokenTypes.struct, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.ORecordChild) {
    pushToken(buffer, range, SemanticTokenTypes.property, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.OType) {
    pushToken(buffer, range, SemanticTokenTypes.type, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.OEnumLiteral && definition.lexerToken?.isLiteral() !== true) {
    pushToken(buffer, range, SemanticTokenTypes.enumMember, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (colorInputs && definition instanceof O.OPort && definition.direction === 'in') {
    pushToken(buffer, range, SemanticTokenTypes.variable, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.OPort) {
    pushToken(buffer, range, SemanticTokenTypes.variable, [...fixedModifiers]);
  } else if (definition instanceof O.OConstant || definition instanceof O.OGeneric) {
    pushToken(buffer, range, SemanticTokenTypes.variable, [...fixedModifiers, SemanticTokenModifiers.readonly]);
  } else if (definition instanceof O.OSignal || definition instanceof O.OVariable || definition instanceof O.OFileVariable) {
    const modifiers = obj instanceof O.OWrite ? [SemanticTokenModifiers.modification] : [];
    pushToken(buffer, range, SemanticTokenTypes.variable, [...fixedModifiers, ...modifiers]);
  } else if (definition instanceof O.OSubprogram) {
    pushToken(buffer, range, SemanticTokenTypes.function, fixedModifiers);
  } else if (definition instanceof O.OAttributeDeclaration) {
    pushToken(buffer, range, SemanticTokenTypes.macro, fixedModifiers);
  } else if (definition instanceof O.OAlias) {
    const aliasDefinition = definition.aliasDefinitions[0];
    if (aliasDefinition) {
      pushCorrectToken(buffer, obj, aliasDefinition, range, fixedModifiers, colorInputs);
    } else {
      pushToken(buffer, range, SemanticTokenTypes.interface, [...fixedModifiers, SemanticTokenModifiers.deprecated]);

    }
  } else {
    console.log(definition.constructor.name);
  }

}

export function semanticToken(linter: VhdlLinter, colorInputs: boolean): SemanticTokens {
  const buffer: BuilderParams[] = [];
  for (const obj of linter.file.objectList) {
    const definition = findDefinition(obj);
    if (definition !== undefined && I.implementsIHasReferenceToken(obj)) {
      pushCorrectToken(buffer, obj, definition, obj.referenceToken.range, [], colorInputs);
    } else if (I.implementsIHasLexerToken(obj)) { // is the definition itself?
      pushCorrectToken(buffer, obj, obj, obj.lexerToken.range, [SemanticTokenModifiers.declaration], colorInputs);
      if (I.implementsIHasEndingLexerToken(obj)) {
        pushCorrectToken(buffer, obj, obj, obj.endingLexerToken.range, [SemanticTokenModifiers.declaration], colorInputs);
      }
    }
    if (obj instanceof O.OArchitecture || obj instanceof O.OInstantiation) {
      pushToken(buffer, obj.entityName.range, SemanticTokenTypes.class, []);
    }
  }
  // Sort tokens
  buffer.sort((a, b) => {
    if (a[0] !== b[0]) {
      return a[0] - b[0];
    }
    return a[1] - b[1];
  });
  const tokensBuilder = new SemanticTokensBuilder();

  for (const row of buffer) {
    tokensBuilder.push(...row);
  }
  return tokensBuilder.build();
}