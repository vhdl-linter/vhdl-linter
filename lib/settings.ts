import { DeepPartial } from 'utility-types';
export interface ISettings {
  ports: {
    outRegex: string;
    inRegex: string;
    enablePortStyle: boolean;
  };
  paths: {
    additional: string[];
    ignoreRegex: string;
  };
  style: {
    preferredLogicTypePort: "unresolved" |  "resolved" | "ignore";
    preferredLogicTypeSignal: "unresolved" | "resolved" | "ignore";
    unusedSignalRegex: string;
    ieeeCasing: 'lowercase' | 'UPPERCASE';
  };
  rules: {
    warnLibrary: boolean;
    warnLogicType: boolean;
    warnMultipleDriver: boolean;
  };
  semanticTokens: boolean;
}
export const defaultSettings: ISettings = {
  ports: {
    outRegex: '^o_',
    inRegex: '^i_',
    enablePortStyle: true,
  },
  paths: {
    additional: [],
    ignoreRegex: ''
  },
  style: {
    preferredLogicTypePort: 'unresolved',
    preferredLogicTypeSignal: 'unresolved',
    unusedSignalRegex: '_unused$',
    ieeeCasing: 'lowercase'
  },
  rules: {
    warnLogicType: true,
    warnLibrary: false,
    warnMultipleDriver: false
  },
  semanticTokens: false
};
export function defaultSettingsGetter() {
  return defaultSettings;
}
export function defaultSettingsWithOverwrite(overwrite?: DeepPartial<ISettings>) {
  const newDefault = Object.assign({}, defaultSettings);

  if (overwrite) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    recursiveObjectAssign(newDefault as any, overwrite);
  }
  return () => defaultSettings;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function recursiveObjectAssign<T extends Record<string, any>>(target: T, source: Partial<T>) {
  Object.keys(source).forEach(key => {
    const s_val = source[key];
    const t_val = target[key];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (target as any)[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
      ? recursiveObjectAssign(t_val, s_val)
      : s_val;
  });
  return target;
}