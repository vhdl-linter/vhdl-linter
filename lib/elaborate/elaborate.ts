import { DiagnosticSeverity } from "vscode-languageserver/node";
import { implementsIHasReference, implementsIHasDefinitions } from "../parser/interfaces";
import { OArchitecture, OEntity, OFile, OPackage, OPackageBody } from "../parser/objects";
import { VhdlLinter } from "../vhdl-linter";
import { elaborateAliases } from "./elaborate-aliases";
import { elaborateAssociations } from "./elaborate-association";
import { elaborateComponents } from "./elaborate-components";
import { elaborateInstantiations } from "./elaborate-instantiations";
import { ElaborateReferences } from "./elaborate-references";
import { ElaborateSelectedNames } from "./elaborate-selected-names";
import { elaborateUseClauses } from "./elaborate-use-clauses";

export class Elaborate {
  file: OFile;
  private constructor(public vhdlLinter: VhdlLinter) {
    this.file = vhdlLinter.file;
  }
  public static async elaborate(vhdlLinter: VhdlLinter) {

    const elaborator = new Elaborate(vhdlLinter);
    await elaborator.elaborateAll();
  }
  public static clear(vhdlLinter: VhdlLinter) {
    for (const obj of vhdlLinter.file.objectList) {
      if (implementsIHasReference(obj)) {
        obj.referenceLinks = [];
        obj.aliasReferences = [];
      }
      if (implementsIHasDefinitions(obj)) {
        obj.definitions = [];
      }
      if (obj instanceof OEntity) {
        obj.correspondingArchitectures = [];
      } else if (obj instanceof OArchitecture) {
        obj.correspondingEntity = undefined;
      } else if (obj instanceof OPackageBody) {
        obj.correspondingPackage = undefined;
      } else if (obj instanceof OPackage) {
        obj.correspondingPackageBodies = [];
      }

    }
  }
  async elaborateAll() {

    await this.vhdlLinter.handleCanceled();

    // const start = Date.now();
    // Map architectures to entity
    for (const architecture of this.file.architectures) {
      if (architecture.entityName === undefined) {
        continue;
      }
      // Find entity first in this file
      let entity = this.file.entities.find(entity => entity.lexerToken.getLText() === architecture.entityName.getLText());
      if (!entity) { // Find entity in all files
        entity = this.vhdlLinter.projectParser.entities.find(entity => entity.lexerToken.getLText() === architecture.entityName.getLText());
      }
      if (entity) {
        architecture.correspondingEntity = entity;
      }

    }
    for (const entity of this.file.entities) {
      entity.correspondingArchitectures = this.vhdlLinter.projectParser.architectures.filter(architecture => entity.lexerToken.getLText() === architecture.entityName.getLText());
    }

    // Map package body to package
    for (const pkg of this.file.packages) {
      if (pkg instanceof OPackageBody) {

        // Find entity first in this file
        let pkgHeader: OPackage | undefined = this.file.packages.find(pkgHeader => pkgHeader instanceof OPackage && pkgHeader.lexerToken.getLText() === pkg.lexerToken.getLText()) as OPackage | undefined;
        if (!pkgHeader) { // Find entity in all files
          pkgHeader = this.vhdlLinter.projectParser.packages.find(pkgHeader => pkgHeader instanceof OPackage && pkgHeader.lexerToken.getLText() === pkg.lexerToken.getLText()) ;
        }
        if (pkgHeader) {
          pkg.correspondingPackage = pkgHeader;
          pkgHeader.correspondingPackageBodies.push(pkg);
        } else {
          this.vhdlLinter.addMessage({
            range: pkg.lexerToken.range,
            severity: DiagnosticSeverity.Warning,
            message: `Can not find package for package body.`
          }, 'elaborate');
        }

      }
    }
    //     console.log(packages);
    elaborateUseClauses(this.file, this.vhdlLinter.projectParser, this.vhdlLinter);
    await this.vhdlLinter.handleCanceled();
    //     console.log(packages);
    // elaborateReferences(this.file);


    // console.log(`elaboration: reads for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();
    elaborateComponents(this.file, this.vhdlLinter.projectParser);
    await this.vhdlLinter.handleCanceled();
    elaborateInstantiations(this.file, this.vhdlLinter.projectParser);

    // console.log(`elaboration: instantiations for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();

    // console.log(`elaboration: components for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();
    ElaborateReferences.elaborate(this.vhdlLinter);
    await this.vhdlLinter.handleCanceled();
    elaborateAssociations(this.file);
    await this.vhdlLinter.handleCanceled();
    elaborateAliases(this.file);
    await this.vhdlLinter.handleCanceled();
    ElaborateSelectedNames.elaborate(this.vhdlLinter);
    // console.log(`elaboration: associations for: ${Date.now() - start} ms.`);
    // start = Date.now();
    await this.vhdlLinter.handleCanceled();

  }







}