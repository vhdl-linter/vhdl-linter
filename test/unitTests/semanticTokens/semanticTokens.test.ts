import { afterAll, beforeAll, beforeEach, expect, jest, test } from '@jest/globals';
import { readdirSync } from 'fs';
import { join } from 'path';
import { pathToFileURL } from 'url';
import { SemanticTokensBuilder } from "vscode-languageserver";
import { Elaborate } from '../../../lib/elaborate/elaborate';
import { semanticToken, semanticTokensLegend } from '../../../lib/languageFeatures/semanticToken';
import { ProjectParser } from '../../../lib/projectParser';
import { defaultSettingsGetter } from '../../../lib/settings';
import { VhdlLinter } from '../../../lib/vhdlLinter';
import { readFileSyncNorm } from "../../readFileSyncNorm";
jest.mock('vscode-languageserver');
let projectParser: ProjectParser;
beforeAll(async () => {
  projectParser = await ProjectParser.create([pathToFileURL(__dirname)], '', defaultSettingsGetter);
});
afterAll(async () => {
  await projectParser.stop();
});
const mockSemanticTokensBuilder = jest.mocked(SemanticTokensBuilder);
beforeEach(() => {
  // Clear all instances and calls to constructor and all methods:
  mockSemanticTokensBuilder.mockClear();
});
const files = readdirSync(__dirname).filter(file => file.endsWith('.vhd'));
const maxLengthTokenTypes = semanticTokensLegend.tokenTypes.reduce((prev, curr) => Math.max(prev, curr.length), 0);
const maxLengthTokenModifiers = semanticTokensLegend.tokenModifiers.reduce((prev, curr) => Math.max(prev, curr.length), 0);
test.each(files.flatMap(file => [[file, true], [file, false]]))('testing semantic tokens for %s direction coloring %p', async (file: string, directionColoring: boolean) => {

  const path = join(__dirname, file);
  const linter = new VhdlLinter(pathToFileURL(path), readFileSyncNorm(path, { encoding: 'utf8' }),
    projectParser, defaultSettingsGetter());
  await Elaborate.elaborate(linter);
  // mock.instances is available with automatic mocks:
  semanticToken(linter, directionColoring);
  const lines = linter.text.split('\n');
  expect(mockSemanticTokensBuilder.mock.instances).toHaveLength(1);
  expect(mockSemanticTokensBuilder.mock.instances[0]?.push.mock.calls.map(([line, char, length, tokenType, tokenModifier]) => {
    const text = lines[line]?.slice(char, char + length);
    const modifiers: (string | undefined)[] = [];
    for (let index = 0; index <= semanticTokensLegend.tokenModifiers.length; index++) {
      if (tokenModifier % 2 === 1) {
        modifiers.push(semanticTokensLegend.tokenModifiers[index]);
      }
      tokenModifier = tokenModifier >> 1;
    }
    return `${line + 1}:${char + 1}`.padStart(10, ' ') + `${(semanticTokensLegend.tokenTypes[tokenType] ?? 'unknown type').padStart(maxLengthTokenTypes, ' ')} ${modifiers.map(modifier => (modifier ?? 'unknown modifier')).join(' ').padStart(maxLengthTokenModifiers, ' ') } ${text ?? 'did not find position in text'}`;
  })).toMatchSnapshot();
});
