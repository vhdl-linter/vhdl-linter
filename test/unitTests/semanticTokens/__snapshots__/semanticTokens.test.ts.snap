// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing semantic tokens for test.vhd 1`] = `
[
  "       1:9        class                ieee",
  "       2:5        class                ieee",
  "      2:10        class                std_logic_1164",
  "       3:5        class                ieee",
  "      3:10        class                numeric_std",
  "       5:8        class                test",
  "       7:5    parameter       readonly a_in",
  "      7:17         enum       readonly std_ulogic",
  "       8:5    parameter                b_out",
  "      8:17         enum       readonly std_ulogic",
  "      10:5        class                test",
  "     12:14        class                arch",
  "     12:22        class                test",
  "     13:10     variable                a",
  "     13:14         type       readonly integer",
  "      18:5    parameter                b_out",
  "     18:14    parameter       readonly a_in",
  "      19:5     variable   modification a",
]
`;

exports[`testing semantic tokens for test_instantiation.vhd 1`] = `
[
  "       1:9        class                ieee",
  "       2:5        class                ieee",
  "      2:10        class                std_logic_1164",
  "       3:5        class                work",
  "      3:10        class                test_pkg",
  "       4:8        class                test_instantiation",
  "      7:14        class                arch",
  "      7:22        class                test_instantiation",
  "      8:10     variable                a",
  "      8:14         enum       readonly std_ulogic",
  "      9:10     variable                b",
  "      9:14         enum       readonly std_ulogic",
  "     11:15        class                entity",
  "     11:22        class                work",
  "     11:27        class                test",
  "      13:5    parameter       readonly a_in",
  "     13:14     variable                a",
  "      14:5    parameter                b_out",
  "     14:14     variable   modification b",
  "     16:11     function                test_procedure",
  "     16:11        class                test_procedure",
]
`;

exports[`testing semantic tokens for test_pkg.vhd 1`] = `
[
  "       1:9        class                ieee",
  "       2:5        class                ieee",
  "      2:10        class                std_logic_1164",
  "       3:5        class                ieee",
  "      3:10        class                numeric_std",
  "       5:9        class                test_pkg",
  "      6:12    parameter       readonly TEST",
  "      6:19         type       readonly integer",
  "       7:8       struct       readonly test_record",
  "       8:5     property       readonly element",
  "      8:14         type       readonly std_ulogic_vector",
  "       9:5     property       readonly another_element",
  "      9:22         type       readonly std_logic_vector",
  "      9:39    parameter       readonly TEST",
  "      10:5     property       readonly banana",
  "     10:13         type       readonly integer",
  "     11:14       struct       readonly test_record",
  "     12:13     function                test_procedure",
  "     12:28    parameter       readonly foo",
  "     12:34         type       readonly integer",
]
`;
